File: Cpp/290_Word_Pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        unordered_map<char, string> pattern_map;
        unordered_map<string, char> word_map;
        string word;
        int pattern_index = 0;
        for(auto& ch : str) {
            if(ch == ' '){
                auto pattern_found = pattern_map.find(pattern[pattern_index]);
                if(pattern_found != pattern_map.end() && pattern_found->second != word) {
                    return false;
                }
                
                auto word_found = word_map.find(word);
                if(word_found != word_map.end() && word_found->second != pattern[pattern_index]) {
                    return false;
                }
                pattern_map[pattern[pattern_index]] = word;
                word_map[word] = pattern[pattern_index];
                word = "";
                ++pattern_index;
            } else {
                word += ch;
            }
        }
        
        auto pattern_found = pattern_map.find(pattern[pattern_index]);
        auto word_found = word_map.find(word);
        return pattern_index == pattern.size() - 1
            && (pattern_found == pattern_map.end() || pattern_found->second == word) 
            && (word_found == word_map.end() || word_found->second == pattern[pattern_index]);
    }
};

File: Kotlin/290_Word_Pattern.kt
class Solution {
    fun wordPattern(pattern: String, s: String): Boolean {
        val words = s.split(" ")
        if(pattern.length != words.size) return false
        
        val patternMap = hashMapOf<Char, String>()
        val wordMap = hashMapOf<String, Char>()
        
        for(i in pattern.indices) {
            val p = pattern[i]
            val word = words[i]
            if(patternMap.containsKey(p)) {
                if(!patternMap[p].equals(word) && wordMap[word] != p) return false
            }
            patternMap.putIfAbsent(p, word)
            wordMap.putIfAbsent(word, p)
            
            if(patternMap.size != wordMap.size) return false
        }
        
        return true
    }
}